packages/router/src/lib/operations/prepare.ts

```diff
@@ -78,14 +80,88 @@ export const prepare = async (
+  let swapPoolIndex: number = -1;
+  config.swapPools.find((pool, index) => {
+    const existSwap =
+      pool.assets.find(
+        (a) =>
+          getAddress(a.assetId) === getAddress(invariantData.sendingAssetId) &&
+          a.chainId === invariantData.sendingChainId,
+      ) &&
+      pool.assets.find(
+        (a) =>
+          getAddress(a.assetId) === getAddress(invariantData.receivingAssetId) &&
+          a.chainId === invariantData.receivingChainId,
+      );
+
+    if (existSwap) {
+      sendingChainIdx = pool.assets.findIndex(
+        (a) =>
+          getAddress(a.assetId) === getAddress(invariantData.sendingAssetId) &&
+          a.chainId === invariantData.sendingChainId,
+      );
+
+      receivingChainIdx = pool.assets.findIndex(
+        (a) =>
+          getAddress(a.assetId) === getAddress(invariantData.receivingAssetId) &&
+          a.chainId === invariantData.receivingChainId,
+      );
+      swapPoolIndex = index;
+    }
+  });
 
-  const outputDecimals = await txService.getDecimalsForAsset(
-    invariantData.receivingChainId,
-    invariantData.receivingAssetId,
+  if (sendingChainIdx == -1 || receivingChainIdx == -1) {
+    throw new SwapInvalid(
+      invariantData.sendingChainId,
+      invariantData.sendingAssetId,
+      invariantData.receivingChainId,
+      invariantData.receivingAssetId,
+      {
+        methodContext,
+        requestContext,
+      },
+    );
+  }
+
+  const swapPool = config.swapPools[swapPoolIndex];
+  const routerBalances = await Promise.all(
+    swapPool.assets.map(async (asset) => {
+      const assetLiquidity = await contractReader.getAssetBalance(asset.assetId, asset.chainId);
+      let assetDecimals = chainData.get(asset.chainId.toString())?.assetId[asset.assetId]?.decimals;
+      if (!assetDecimals) {
+        assetDecimals = await txService.getDecimalsForAsset(asset.chainId, asset.assetId);
+      }
+      // convert asset liquidity into 18 decimal value.
+      const res = assetLiquidity.mul(BigNumber.from(10).pow(18 - assetDecimals));
+      return res;
+    }),
   );
 
-  let receiverAmount = await getReceiverAmount(senderAmount, inputDecimals, outputDecimals);
+  let inputDecimals = chainData.get(invariantData.sendingChainId.toString())?.assetId[invariantData.sendingAssetId]
+    ?.decimals;
+  if (!inputDecimals) {
+    inputDecimals = await txService.getDecimalsForAsset(invariantData.sendingChainId, invariantData.sendingAssetId);
+  }
+  let outputDecimals = chainData.get(invariantData.receivingChainId.toString())?.assetId[invariantData.receivingAssetId]
+    ?.decimals;
+  if (!outputDecimals) {
+    outputDecimals = await txService.getDecimalsForAsset(
+      invariantData.receivingChainId,
+      invariantData.receivingAssetId,
+    );
+  }
+
+  let receiverAmount = await getReceiverAmount(
+    senderAmount,
+    inputDecimals,
+    outputDecimals,
+    routerBalances,
+    sendingChainIdx,
+    receivingChainIdx,
+    config.maxPriceImpact,
+  );
```

Comment:
if the `PriceImpactTooHigh` error is thrown, we should mark that error as cancellable, so that the router will cancel the users transaction automatically if it cannot prepare

Reply:
ok. I got you